### PR TITLE
refactorization of book class

### PR DESCRIPTION
**Pull request by dendorferpatrick**

### CODE REVIEW

File: library/model/book.py [hunk 1]
```diff
@@ -168,3 +145,46 @@ def serialize(self, book: Book, format: str):
             return et.tostring(book_info, encoding="Unicode")
         else:
             raise ValueError(format)
+
+class BookType():
```
Comment (WilliamLindskog) [inline library/model/book.py:149]: good use of abstract class

File: library/model/book.py [hunk 2]
```diff
@@ -168,3 +145,46 @@ def serialize(self, book: Book, format: str):
...
+
+    @abstractmethod
+    def get_approximate_duration(self, book: Book):
+        pass
+    
+    def get_weekly_fees(self):
+        return self.weekly_fees
+
+class PaperType(BookType):
+    book_type: str = "paper"
```
Comment (WilliamLindskog) [inline library/model/book.py:164]: can this be reached from outside of the class ?

File: library/model/book.py [hunk 3]
```diff
@@ -20,130 +21,106 @@ class Book:
...
-        if self._book_type == "Paper":
-            return self.existing_items - self.borrowed_items > 0
-        elif self._book_type == "Electronic":
-            return True
-        elif self._book_type == "Audio":
-            return True
+        self.borrowed_items = borrowed_items
+        self.in_stock_items = self.existing_items - self.borrowed_items
+
+    def init_existing_books(self):
```
Comment (Rene36) [inline library/model/book.py:51]: Add a short function description. It is at first glance hard to understand that you calculate the total number of books.

File: library/model/book.py [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
 from library.model.author import Author
 from library.model.genre import Genre
 from library.model.publisher import Publisher
 import xml.etree.ElementTree as et
 
 from library.persistence.storage import LibraryRepository
 
-
-class Book:
+class BookSerialNumber:
```
Comment (WilliamLindskog) [inline library/model/book.py:12]: should book serial number class include authors and title?

File: library/model/book.py [hunk 5]
```diff
@@ -20,130 +21,106 @@ class Book:
...
-            return self.pages * 3 * 60
-        elif self._book_type == "Electronic":
-            return self.pages * 5 * 60
-        elif self._book_type == "Audio":
-            return self.duration
-        else:
-            raise AttributeError("No such book type...")
 
+    def can_borrow(self) -> bool:
+        return self._book_type.can_borrow(self) and len(self.in_stock_list) > 0
```
Comment (WilliamLindskog) [inline library/model/book.py:71]: Maybe try to use one "can_borrow" function instead of 2

File: library/model/book.py [hunk 6]
```diff
@@ -20,130 +21,106 @@ class Book:
...
-        elif self._book_type == "Audio":
-            return self.duration
-        else:
-            raise AttributeError("No such book type...")
 
+    def can_borrow(self) -> bool:
+        return self._book_type.can_borrow(self) and len(self.in_stock_list) > 0 
+       
+    def get_approximate_duration(self) -> int:
+        return self._book_type.get_approximate_duration(self)
```
Comment (WilliamLindskog) [inline library/model/book.py:74]: can we only use 1 function instead of 2?

File: library/model/book.py [hunk 7]
```diff
@@ -20,130 +21,106 @@ class Book:
...
-        elif self._book_type == "Electronic":
-            return 2
-        elif self._book_type == "Audio":
-            return 2
-        else:
-            raise AttributeError("No such book type...")
+        return self._book_type.get_weekly_fees()
+
+    def check_books_stock_consistency(self):
+        assert len(self.borrowed_books_list) + self.in_stock_list)== self.existing_items, "Stock not consistent!!!"
```
Comment (Rene36) [inline library/model/book.py:80]: Do not use assertions in production code.

File: library/model/book.py [hunk 8]
```diff
@@ -20,130 +21,106 @@ class Book:
...
-        if self.can_borrow():
-            if self._book_type == "Paper":
-                self.borrowed_items += 1
+
+        if self.can_borrow(): 
             LibraryRepository.update_book(self)
-            borrowed_book = BorrowedBook.from_book(self)
-            borrowed_book.due_date = datetime.now() + timedelta(days=7)
-            borrowed_book.current_fee = self.get_weekly_fee()
+            borrowed_book = self.get_in_stock_books()[0]
```
Comment (Rene36) [inline library/model/book.py:86]: Why take the first list entry of all books in stock and not the 2nd or 3rd?

File: library/model/book.py [hunk 9]
```diff
@@ -20,130 +21,106 @@ class Book:
...
     def renew_rental(self) -> "BorrowedBook":
-        self.due_date += timedelta(days=7)
+        self.due_date += timedelta(days=self._renew_rental_data)
         self.current_fee += self.get_weekly_fee()
         return self
+    def set_current_fee(self, current_fee): 
+        self.current_fee = current_fee
+    
+    def borrowed_book(self):
+        self._is_borrowed = False
```
Comment (Rene36) [inline library/model/book.py:113]: From the method name not clear that it sets the internal class variable _is_borrowed to False.

**Review by Rene36 — COMMENTED** [review]
See comments

**Review by WilliamLindskog — COMMENTED** [review]
final submission

**Pull request closed without merging**